File: src/main/java/org/snakeyaml/engine/api/DumpSettings.java
/**
 * Copyright (c) 2018, http://www.snakeyaml.org
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.snakeyaml.engine.api;

import java.util.Map;
import java.util.Optional;

import org.snakeyaml.engine.common.FlowStyle;
import org.snakeyaml.engine.common.ScalarStyle;
import org.snakeyaml.engine.common.SpecVersion;
import org.snakeyaml.engine.nodes.Tag;
import org.snakeyaml.engine.resolver.ScalarResolver;
import org.snakeyaml.engine.serializer.AnchorGenerator;

/**
 * Fine tuning serializing/dumping
 * TODO add JavaDoc for all methods
 */
public final class DumpSettings {
    private String label;
    private boolean explicitStart;
    private boolean explicitEnd;
    private Optional<Tag> explicitRootTag;
    private AnchorGenerator anchorGenerator;
    private Optional<SpecVersion> specVersion;
    private Map<String, String> useTags;
    private ScalarResolver scalarResolver;
    private FlowStyle defaultFlowStyle;
    private ScalarStyle defaultScalarStyle;

    //emitter
    private Optional<Tag> rootTag;
    private boolean canonical;
    private boolean prettyFlow;
    private boolean useUnicodeEncoding;
    private int indent;
    private int indicatorIndent;
    private int width;
    private String bestLineBreak;
    private boolean splitLines;


    DumpSettings(String label, boolean explicitStart, boolean explicitEnd, Optional<Tag> explicitRootTag,
                 AnchorGenerator anchorGenerator, Optional<SpecVersion> specVersion, Map<String, String> useTags,
                 ScalarResolver scalarResolver, FlowStyle defaultFlowStyle, ScalarStyle defaultScalarStyle,
                 //emitter
                 Optional<Tag> rootTag, boolean canonical, boolean prettyFlow, boolean useUnicodeEncoding,
                 int indent, int indicatorIndent, int width, String bestLineBreak, boolean splitLines
    ) {
        this.label = label;
        this.explicitStart = explicitStart;
        this.explicitEnd = explicitEnd;
        this.explicitRootTag = explicitRootTag;
        this.anchorGenerator = anchorGenerator;
        this.specVersion = specVersion;
        this.useTags = useTags;
        this.scalarResolver = scalarResolver;
        this.defaultFlowStyle = defaultFlowStyle;
        this.defaultScalarStyle = defaultScalarStyle;
        this.rootTag = rootTag;
        this.canonical = canonical;
        this.prettyFlow = prettyFlow;
        this.useUnicodeEncoding = useUnicodeEncoding;
        this.indent = indent;
        this.indicatorIndent = indicatorIndent;
        this.width = width;
        this.bestLineBreak = bestLineBreak;
        this.splitLines = splitLines;
    }


    public FlowStyle getDefaultFlowStyle() {
        return defaultFlowStyle;
    }

    public ScalarStyle getDefaultScalarStyle() {
        return defaultScalarStyle;
    }

    public String getLabel() {
        return label;
    }

    public boolean isExplicitStart() {
        return explicitStart;
    }

    public AnchorGenerator getAnchorGenerator() {
        return anchorGenerator;
    }

    public ScalarResolver getScalarResolver() {
        return scalarResolver;
    }

    public boolean isExplicitEnd() {
        return explicitEnd;
    }

    public Optional<Tag> getExplicitRootTag() {
        return explicitRootTag;
    }

    public Optional<SpecVersion> getSpecVersion() {
        return specVersion;
    }

    public Map<String, String> getUseTags() {
        return useTags;
    }

    public Optional<Tag> getRootTag() {
        return rootTag;
    }

    public boolean isCanonical() {
        return canonical;
    }

    public boolean isPrettyFlow() {
        return prettyFlow;
    }

    public boolean isUseUnicodeEncoding() {
        return useUnicodeEncoding;
    }

    public int getIndent() {
        return indent;
    }

    public int getIndicatorIndent() {
        return indicatorIndent;
    }

    public int getWidth() {
        return width;
    }

    public String getBestLineBreak() {
        return bestLineBreak;
    }

    public boolean isSplitLines() {
        return splitLines;
    }

}

File: src/commonMain/kotlin/org/snakeyaml/engine/v2/api/DumpSettings.kt
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.v2.api

import org.snakeyaml.engine.v2.common.FlowStyle
import org.snakeyaml.engine.v2.common.NonPrintableStyle
import org.snakeyaml.engine.v2.common.ScalarStyle
import org.snakeyaml.engine.v2.common.SpecVersion
import org.snakeyaml.engine.v2.nodes.Tag
import org.snakeyaml.engine.v2.schema.Schema
import org.snakeyaml.engine.v2.serializer.AnchorGenerator
import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

/**
 * Immutable configuration for serialisation. Description for all the fields can be found in the
 * builder
 */
class DumpSettings internal constructor(
    val isExplicitStart: Boolean,
    val isExplicitEnd: Boolean,
    @JvmField val explicitRootTag: Tag?,
    @JvmField val anchorGenerator: AnchorGenerator,
    @JvmField val yamlDirective: SpecVersion?,
    @JvmField val tagDirective: Map<String, String>,
    @JvmField val defaultFlowStyle: FlowStyle,
    @JvmField val defaultScalarStyle: ScalarStyle,
    @JvmField val nonPrintableStyle: NonPrintableStyle,
    @JvmField val schema: Schema,
    // emitter
    val isCanonical: Boolean,
    val isMultiLineFlow: Boolean,
    val isUseUnicodeEncoding: Boolean,
    @JvmField val indent: Int,
    @JvmField val indicatorIndent: Int,
    @JvmField val width: Int,
    @JvmField val bestLineBreak: String,
    val isSplitLines: Boolean,
    @JvmField val maxSimpleKeyLength: Int,
    // general
    private val customProperties: Map<SettingKey, Any>,
    @JvmField val indentWithIndicator: Boolean,
    @JvmField val dumpComments: Boolean,
) {

    fun getCustomProperty(key: SettingKey): Any? = customProperties[key]

    companion object {
        @JvmStatic
        fun builder(): DumpSettingsBuilder = DumpSettingsBuilder()
    }
}
